### ISSUE TITLE
Extension Naming Validation

### ISSUE THREAD

**Issue by MB1211** [cloudevents/sdk-java#388]
Using JsonFormat, I am getting exceptions thrown for any extension name that is not strictly alphanumeric. I don't think this issue is necessarily specific to JsonFormat though

The spec states "CloudEvents attribute names MUST consist of lower-case letters ('a' to 'z') or digits ('0' to '9') from the ASCII character set. Attribute names SHOULD be descriptive and terse and SHOULD NOT exceed 20 characters in length."

But the code BaseCloudEventBuilder to validate the extension name (which references https://github.com/cloudevents/spec/blob/master/spec.md#attribute-naming-convention) seems to be misinterpreting this. It interprets it as MUST ONLY, because the method to validate characters is the following:

```
private static boolean isValidChar(char c) {
        return (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9');
    }
```
Unless I'm mistaken, this is not the intent, and a fairly problematic bug. It's not really possible to be descriptive with no way of delimiting words. Elsewhere in the spec, and in the code itself there are mentions of extensions with -'s and _'s so these at the very least should be allowed (ce-, ce_).

**Comment by slinkydeveloper** [comment]
Hi, this was fixed in 2.1 version of the sdk https://github.com/cloudevents/sdk-java/pull/366

**Comment by MB1211** [comment]
This is a similar issue but it is not the same. The referenced PR only addresses the validation of the length as being SHOULD NOT and not MUST NOT. This issue is about the validity of the characters. Please reopen
